### DISCUSSION TITLE
Pop-out Virtual Flag Panel

### DISCUSSION THREAD

**Discussion by JJWatMyself** [f1multiviewer/issue-tracker#76] [Feature requests] [answered]
The feature request is for a pop-out window, that can either be a window or full screen virtual flag panel.  This would show the changes in flag status for all parts of the screen, similar to what is displayed on the corner of the screen for the race.  Designing the Virtual Flag Panel to display in a similar fashion to the real flag panels on the track would provide a driver view like experience.  The virtual flag panel would have the added benefit that if the user has smart lights that can adjust color based on the computer display (e.g. Philips Hue), then the virtual lights would automatically synchronize with the virtual flag panel.

Linked is a video of the track side flag panels.

https://youtu.be/-Bq1eD7g-MA

**Comment by JJWatMyself** [comment]
Flag panel control example.

https://www.alamy.com/stock-photo-f1-marshalling-system-electronic-flag-panel-control-pictured-at-the-55511590.html

**Comment by JJWatMyself** [comment]
If this were to be like an actual track Electronic Flag Panel, then the model from the video has 50 horizontal LEDs and 40 vertical LEDs. There are a number of panel models and provided as a link in the next post.

![image](https://user-images.githubusercontent.com/22590445/176052663-15f2ff08-2a70-4497-98af-b499635b64e6.png)

**Comment by JJWatMyself** [comment]
Flag modes for an official EM Motorsport Electronic Flag Panel are:

Panel off / no flag / state when flashing on-off
![image](https://user-images.githubusercontent.com/22590445/176052960-fc8d70b0-4e0b-4146-9d06-f348edfeeae4.png)

Slippery Surface
![image](https://user-images.githubusercontent.com/22590445/176052800-5eecfb2c-d34b-45b6-80ed-638d8de7dc88.png)

Yellow flag (flashing on-off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176052880-0ea0c911-2873-4f35-88fd-4e9b53b2e84b.png)

Double yellow (rapid flashing opposite diagonal images, 2020 spec states 2-4 Hz, video is 3 Hz)
![image](https://user-images.githubusercontent.com/22590445/176053016-6eeb97ea-3f12-45e9-8c05-93bd919db60a.png)
![image](https://user-images.githubusercontent.com/22590445/176053066-27a353e1-9783-4810-9554-2dde4bdaff07.png)

Full course yellow flag
![image](https://user-images.githubusercontent.com/22590445/176063845-80fbd221-6f19-4691-b85a-159a3371958f.png)

Blue flag (flashing on-off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176052750-4ab03752-a96b-4015-8322-0fdf63a51f58.png)

Green flag (flashing on-off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176053177-a6941e15-ed92-47da-bdca-e61471efeaa5.png)

Red flag (flashing on-off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176053215-8162e6ba-9036-411e-b607-71700912ce54.png)

Safety Car Flag (flashing two images at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176053261-a9708092-6cd5-41c4-b6d3-f125ec38a35d.png)
![image](https://user-images.githubusercontent.com/22590445/176053359-a076335c-77c2-484c-8eb9-904c124901e5.png)

Virtual Safety Car (flashing two images at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176053457-d9c49aca-286e-4296-8908-bec5e3b5a324.png)
![image](https://user-images.githubusercontent.com/22590445/176053477-87e81ee0-86cd-46d2-9121-d5be12b0c0f3.png)

White flag (flashing on-off at 2 Hz) interestingly we see distinct colors on each border (yellow, magenta, red, cyan) and reminds me of DLP 5-color wheel which uses, red, green, blue, cyan, yellow
![image](https://user-images.githubusercontent.com/22590445/176053521-b001cf27-7c49-49df-be40-0f3b7340431e.png)

Standing start flag (pulse flashing long on-short off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176054104-cfb868e6-bc61-4f46-893c-01f133777a3e.png)

Rolling start flag (pulse flashing long on-short off at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176054161-8a4388d2-b5ac-435b-9aac-d2edfe04b18e.png)

Chequered flag (flashing two images at 2 Hz)
![image](https://user-images.githubusercontent.com/22590445/176054208-9e6a5c63-2643-443a-9269-e99d4700c3b6.png)
![image](https://user-images.githubusercontent.com/22590445/176054267-7f8de788-df99-4e37-884c-d59f60175a8b.png)

Pit entry flag (open)
![image](https://user-images.githubusercontent.com/22590445/176063998-5bf776fe-9bb2-42ab-a84f-e035b2ade4f8.png)

Pit entry closed flag
![image](https://user-images.githubusercontent.com/22590445/176058067-683f1746-2763-4fe3-9dc0-debdcc625b87.png)

Wet track/rain function, blue animated droplets, not referenced in the current FIA documents, although evidenced on Monaco 2022 video footage. Hold 'Slippery button for 5 seconds' although my stop watch shows it is held for 4 seconds, not 5.
![unknown](https://user-images.githubusercontent.com/22590445/188865774-5713318b-bfb5-486e-8cfd-cb666fac308c.png)

Example of meatball flag:
![image](https://user-images.githubusercontent.com/22590445/190012054-dd7c456e-930b-4c27-97b2-ec972c698105.png)

Example of black and white flag (unsportsmanlike warning before black flag disqualification):
![image](https://user-images.githubusercontent.com/22590445/190012436-a8ed6283-ddbc-4645-b188-1d22fb9e8c0b.png)

Worth noting that at least one manufacturer has a mini panel on the rear that shows the same as the front. Ref Monaco 2022 video footage for red flag example.  This is referenced in the spec documents, however worth noting there are areas where the rear is covered with a sponsor poster, ref Dutch GP 2022.
![unknown](https://user-images.githubusercontent.com/22590445/188867207-9c99db38-7372-4c85-864e-447c0f509ab6.png)
![image](https://user-images.githubusercontent.com/22590445/188944116-74e8b724-312b-4bba-b1c7-a71a91b102a2.png)

Picture of portable 'Bluetooth' controller by Alkamel, with mini DigiFlag LED panel built-in:
![image](https://user-images.githubusercontent.com/22590445/190013533-2554d1c1-8a8d-47d3-9e62-62c30e46644b.png)
![image](https://user-images.githubusercontent.com/22590445/190013642-92159de0-9b0e-4590-9129-bb14acc8a592.png)
![image](https://user-images.githubusercontent.com/22590445/190013671-2bd977b2-467c-4276-b82d-4d5e342e32be.png)

Picture of wired controller:
![image](https://user-images.githubusercontent.com/22590445/190013979-bb2799fb-ac0f-4fc5-8501-f42f52ba9f76.png)
Close-up image:
![image](https://user-images.githubusercontent.com/22590445/190015528-daf38598-e9e0-4d58-98a4-c7e06019d38c.png)

To give you an idea of scale of the panel:
![image](https://user-images.githubusercontent.com/22590445/190014478-904cb63d-fb89-4f41-8824-198f8172ffaf.png)

Note, the following are missing

black flag

Introduction of DigiFlag April 2008:
https://www.pitpass.com/public/print_article.php?fes_art_id=34527

Introduction of VSC Jan 2015, due to 2014 Japanese GP crash:
https://www.speedcafe.com/2015/01/09/fia-reveals-f1-virtual-safety-car-procedure/

Page 25 here has a full FIA spec: https://www.fia.com/sites/default/files/fia_standard_3504-2019_light_panels_en.pdf

Version 2 spec March 2020 here has frequency ranges https://www.fia.com/sites/default/files/fia_standard_3504-2019_light_panels_en_v2.pdf

2021-03-21 all official panel part numbers and their manufacturers: https://www.fia.com/sites/default/files/technical_list_ndeg81_-_light_panels_homologated_manufacturers_-_fia_standard_3504-2019_1.pdf

Informative article: https://www.fia.com/sites/default/files/auto_medical_19.pdf
(Can open other issues by updating the number to the issue number, e.g. 10)

Amazing resource here: https://www.racecar-engineering.com/articles/track-marshalling-systems-em-motorsport/

More recent news and that panels are moved between circuits by the race promoter, e.g. Formula 1, and states FIA Grade 1 requires T1-spec homologated light panels.

https://www.crash.net/motogp/news/949213/1/light-panels-be-made-mandatory-motogp-and-f1-2022

May 2022 IndyCar adopted the standard. Hardware appears to be different to the F1 hardware, albeit made by EM Motorsport as far as I can tell from this article:  https://www.pmw-magazine.com/news/race-series-news/indycar-updates-to-electronic-marshalling-system-with-digital-flag-panels.html

Sept 13 2022 Tech Talk 'F1 Flags explained' by Albert Fabrega
https://www.youtube.com/watch?v=_4UusnCaB6s

**Comment by JJWatMyself** [comment]
Nice image of an Electronic Flag Panel.
![homeSlider-T1_1570611029](https://user-images.githubusercontent.com/22590445/176057421-68431b80-a765-4cda-ae6b-e7c0b3e122eb.jpg)

**Comment by JJWatMyself** [comment]
It appears that these are also known as FIA DigiFlags
![FIA_Digiflag_SC](https://user-images.githubusercontent.com/22590445/176058956-bb021380-2f4d-4f05-bccf-cb80d6813232.png)

https://commons.wikimedia.org/w/index.php?search=digiflags&title=Special:MediaSearch&go=Go&type=image

**Comment by JJWatMyself** [comment]
A hi-res image of a relatively modern panel.
![5304017_HiRes](https://user-images.githubusercontent.com/22590445/176059783-09244275-79fc-46ce-b187-29862d89118b.jpg)

**Reply by JJWatMyself** [reply]
Perfect hardware for this might be the Pixoo 64.

https://youtu.be/zm0IpeetvvY

https://www.divoom.com/products/pixoo-64?variant=39473700110454

![image](https://user-images.githubusercontent.com/22590445/177021370-f9d595ff-0abf-470a-acb9-c40fd8fc5b29.png)

**Reply by JJWatMyself** [reply]
Pixoo 64 API to play GIF here:

http://doc.divoom-gz.com/web/#/12?page_id=195

**Comment by JJWatMyself** [comment]
I created some FIA spec DigiFlag GIF files last night. These should run on the Pixoo 64.

First test files with issues
![DigiFlag - Slippery 64x64](https://user-images.githubusercontent.com/22590445/177052928-4c9fd63e-0c5b-4baa-85c1-9a6a5b4c5c5b.gif)    ![DigiFlag - Double Yellow 64x64](https://user-images.githubusercontent.com/22590445/177052929-4ad617fe-323b-45ee-ac09-4203f95912d9.gif)    ![DigiFlag - Yellow 64x64](https://user-images.githubusercontent.com/22590445/177052930-5e6e1882-6e36-4044-b582-2c585a46ae78.gif)    ![DigiFlag - White 64x64](https://user-images.githubusercontent.com/22590445/177052931-fc30a4fd-813f-41be-a8c3-ed556ba7f20a.gif)

![DigiFlag - Red 64x64](https://user-images.githubusercontent.com/22590445/177052932-0e18f1ca-6fc4-4518-85e9-64c07a4cbd88.gif)    ![DigiFlag - Blue 64x64](https://user-images.githubusercontent.com/22590445/177052933-3cbcd5cb-0381-494a-bf59-cdbecc598771.gif)    ![DigiFlag - Green 64x64](https://user-images.githubusercontent.com/22590445/177052934-6b81ce04-12aa-400b-8f94-89b41bb4e270.gif)    ![DigiFlag - Chequered 64x64](https://user-images.githubusercontent.com/22590445/177058196-8cfa5b2e-532f-48ab-a3e2-d41460f547aa.gif)

![DigiFlag - Pit Entry Closed 64x64](https://user-images.githubusercontent.com/22590445/177058473-f8099735-70aa-464e-a13f-add19b36044f.gif)    ![DigiFlag - Pit Entry 64x64](https://user-images.githubusercontent.com/22590445/177058693-24ddf0ce-d300-4ad4-9506-77ebadfcec4d.gif)

Second test files without issues (requires black pixels)
![DigiFlag - Double Yellow 64x64-black](https://user-images.githubusercontent.com/22590445/178840774-511bb329-64b7-4fb4-b1da-2bd63373b293.gif)    ![DigiFlag - Chequered 64x64-black](https://user-images.githubusercontent.com/22590445/178841126-40957415-c097-4835-961f-9d67a1832deb.gif)    ![DigiFlag - Red 64x64-black](https://user-images.githubusercontent.com/22590445/178846253-af2c5aa1-a02e-46d0-bfb9-1700834d75ca.gif)    ![DigiFlag - Slippery 64x64-black](https://user-images.githubusercontent.com/22590445/178846254-501aa024-ee84-4f8c-9d5b-74641fc6d1a2.gif)    ![DigiFlag - White 64x64-black](https://user-images.githubusercontent.com/22590445/178846256-4d55a9ef-7de0-4704-b8ac-4c9fc030d7ee.gif)    ![DigiFlag - Yellow 64x64-black](https://user-images.githubusercontent.com/22590445/178846257-c4cace7b-e8dd-406c-bf5e-e513f80199e0.gif)    ![DigiFlag - Blue 64x64-black](https://user-images.githubusercontent.com/22590445/178846259-707b0171-8439-4939-88a7-cc9fbc4cdb04.gif)    ![DigiFlag - Green 64x64-black](https://user-images.githubusercontent.com/22590445/178846261-dda96a56-7b06-4075-976b-dfc4bb035743.gif)    ![DigiFlag - Pit Entry 64x64-black](https://user-images.githubusercontent.com/22590445/178846262-ce32cf94-e376-4e60-a65a-9754260fdc78.gif)    ![DigiFlag - Pit Entry Closed 64x64-black](https://user-images.githubusercontent.com/22590445/178846265-bfa5aed7-6724-4af5-829b-15f04f6796be.gif)

Third test files to address speed of flashing and add some others.

![red](https://user-images.githubusercontent.com/22590445/193566621-9f343313-1199-43fa-b867-6c39c243b604.gif) ![sc](https://user-images.githubusercontent.com/22590445/193566622-488d60ef-7824-4fd4-8d0c-b7953a32889b.gif) ![slippery](https://user-images.githubusercontent.com/22590445/193566624-be23a2e4-ae01-4a11-8ad0-65e63b535652.gif) ![void](https://user-images.githubusercontent.com/22590445/193566625-23a5faf5-0caa-41a9-975c-2ee802780c93.gif) ![vsc](https://user-images.githubusercontent.com/22590445/193566626-934d5779-9b06-47e3-a9cc-85729cf1242a.gif) ![white](https://user-images.githubusercontent.com/22590445/193566627-d4ead4a8-1c60-4af5-b498-0bc50f81794d.gif) ![yellow](https://user-images.githubusercontent.com/22590445/193566630-051fcaf4-4935-461a-85b3-da67ec0a67eb.gif) ![blackandwhite](https://user-images.githubusercontent.com/22590445/193566631-c0f3cb9c-cd14-4db7-81d0-95b471776659.gif) ![blue](https://user-images.githubusercontent.com/22590445/193566633-225c2685-2ed7-4058-82a1-197571141189.gif) ![chequered](https://user-images.githubusercontent.com/22590445/193566634-a3567867-0434-4b9c-bb8f-1a4db7e0231d.gif) ![dyellow](https://user-images.githubusercontent.com/22590445/193566635-2c01d6c4-c541-4fe5-93aa-a95bcf064a20.gif) ![green](https://user-images.githubusercontent.com/22590445/193566636-ea386d47-434b-4cf6-accd-dc84297762b0.gif) ![mec](https://user-images.githubusercontent.com/22590445/193566639-63ad05e7-1c5d-481d-b266-12c7c2bf8815.gif) ![pitclosed](https://user-images.githubusercontent.com/22590445/193566640-54256e02-6960-4d60-95fb-93d9b79b91e7.gif) ![pitentry](https://user-images.githubusercontent.com/22590445/193566642-be763b0f-aa18-4729-9594-d9d3e3e39acf.gif) ![rain](https://user-images.githubusercontent.com/22590445/193566644-6b5d4897-a325-43b5-9ceb-35ecd5175458.gif)

MultiViewer logo versions:

![mvlogo2](https://user-images.githubusercontent.com/22590445/193677323-fbbf62c3-00cb-45c4-9fc1-12e3c7a07263.gif)
![mvlogo](https://user-images.githubusercontent.com/22590445/193677324-f76d819c-827e-4276-8664-8b894ff4e218.gif)

This is the site used to create:
https://www.piskelapp.com/p/create/sprite


EM Motorsport GIF Files. There will likely be some edits after reviewing some recent video footage:
![chequered](https://user-images.githubusercontent.com/22590445/188344762-6e863831-d9eb-40e5-91e0-1e0f3bc39c66.gif)
![dyellow](https://user-images.githubusercontent.com/22590445/188344763-a3497307-981a-4bd6-a033-8c910f129635.gif)
![rs](https://user-images.githubusercontent.com/22590445/188344767-1f81c10f-8633-445b-a5ad-24072c9e023b.gif)
![ss](https://user-images.githubusercontent.com/22590445/188868099-a1281d94-2fc8-4ad2-9172-0faeecc1764f.gif)
![slippery](https://user-images.githubusercontent.com/22590445/188344771-1ec59090-2fa9-43b4-830c-a36e07cfa249.gif)
![void](https://user-images.githubusercontent.com/22590445/188344774-34fffe38-1908-4b6e-aef0-2ed4278d6aee.gif)
![white](https://user-images.githubusercontent.com/22590445/188344779-e83d6bc5-567b-45a5-a5a8-f1e675a29d62.gif)

After review of on board cameras Dutch and British GP 2022
![Yellow](https://user-images.githubusercontent.com/22590445/188508727-4e50cb30-4bc9-4af9-8b6c-c4cff75842bf.gif)
![Virtual Safety Car](https://user-images.githubusercontent.com/22590445/188508730-de7a60b1-e46a-425f-b177-1374b8ec3fab.gif)
![Safety Car](https://user-images.githubusercontent.com/22590445/188508731-47d73777-d20f-41c9-8706-eb5fe9c5e593.gif)
![Green](https://user-images.githubusercontent.com/22590445/188508734-976fe771-a2d1-4dc4-9149-7fe8697c070d.gif)
![Red](https://user-images.githubusercontent.com/22590445/188509826-673e5c2d-209d-42b5-a614-ecd7c34c507a.gif)
![Blue](https://user-images.githubusercontent.com/22590445/188509827-d8b150ab-4d81-4282-ab9c-25f846b4e231.gif)

**Comment by Psybertr0n** [comment]
Having that external pixel display to show some extra information is a great idea but I feel that lone display in our setup cannot realistically provide all the information with the full context that the electronic flags all around the track do. For example, it makes sense for the pitlane entry panel to show only an arrow (lane open) or a red X (lane closed) all the time because it is situated at pit entry and has that context. Similarly, a double yellow flag by itself doesn't mean anything but when it is placed physically at a mini-sector of the track it means there's double yellows at that specific mini-sector. 

I don't think a single pixel panel with its limited real estate will be able to show all that information, especially stuff like blue flags and white flags that are shown at specific parts of the track. It may be more beneficial to divide the panel into 4 quadrants and have each quadrant display the status of only the most relevant flags (track status, pitlane status and so on) but I don't know if there's enough resolution in the pixel panel to display smaller text to label those 4 flags.

However a virtual panel in the form of a dedicated window in our OS won't have these limitations. It can display the key information of the flags that would be relevant to us viewers. Alpha Tauri's pitwall pictured below can be taken as inspiration. The engineer has 4 flag statuses displayed to them in their pitwall screen. 

![chrome_2022-07-04_13-59-30](https://user-images.githubusercontent.com/16277629/177117736-ffb2eec5-a9a5-43bf-b4da-d906b8999ed1.jpg)

Cropped and zoomed-in view of the flags:
![ImageGlass_2022-07-04_14-15-08](https://user-images.githubusercontent.com/16277629/177117839-2ab66003-85c9-4717-bf9b-52cba16b3c88.png)

I think we can have a dedicated popout window to show these same statuses because these are the only relevant ones to keep an eye on. I can understand what the middle 2 flags are: track status and pit entry+exit flags. I'm not entirely sure about the other two. The track status flag can show green, red, yellow or double yellow with sector label, VSC or SC. The sector label can be extracted out of the RCM JSON stream.

**Comment by JJWatMyself** [comment]
Link to video of working GIF files playing on the Pixoo 64.

https://discord.com/channels/982375740764983366/984804585317208084/996913148990861414

**Comment by JJWatMyself** [comment, marked as answer]
Amazing work on window for virtual electronic flag panel by LapsTimeOFF found here:

https://github.com/LapsTimeOFF/DigiFlag_F1MV

**Reply by LapsTimeOFF** [reply]
thanks lol

**Comment by JJWatMyself** [comment]
Rain, rain rain.  Thanks to TheBaconCat it caused me to question the frames that make up the rain animation.  So back to the drawing board and confirming and concurring with TheBaconCat that there are four frames, running at exactly 4 FPS.  Here are extracted files and a GIF.

![scene00019](https://user-images.githubusercontent.com/22590445/196008091-eb68ea45-00cd-4cb9-a77d-acb463718576.png)
![scene00020](https://user-images.githubusercontent.com/22590445/196008092-58e2d5bb-6d33-4ac7-884f-75206a48ec9c.png)
![scene00021](https://user-images.githubusercontent.com/22590445/196008093-3b31b4f1-0bf1-407d-8f3d-f79fdbe33705.png)
![scene00022](https://user-images.githubusercontent.com/22590445/196008096-dd0b6f57-400f-4657-abdb-2d9410ab1d4a.png)

![ezgif-2-11dd9badcf](https://user-images.githubusercontent.com/22590445/196008103-c9ca2ea2-ce9c-4f98-8c51-d73d13582bdf.gif)

**Comment by JJWatMyself** [comment]
Amazing work by Readis here:

![image 1](https://user-images.githubusercontent.com/22590445/202695015-df4acc6b-81d0-4ea3-a652-e433cbcc00d4.png)

(https://discord.com/channels/982375740764983366/986304513990336573/1042836838441824376)
